pallets/dapp-staking-v3/src/lib.rs

```diff
@@ -1695,24 +1718,26 @@ pub mod pallet {
                 tier_id.saturating_inc();
             }
 
+            // TODO: what if multiple dApps satisfy the tier entry threshold but there's not enough slots to accomodate them all?
+
             // 4.
-            // Sort by dApp ID, in ascending order (unstable sort should be faster, and stability is guaranteed due to lack of duplicated Ids).
-            // TODO & Idea: perhaps use BTreeMap instead? It will "sort" automatically based on dApp Id, and we can efficiently remove entries,
-            // reducing PoV size step by step.
-            // It's a trade-off between speed and PoV size. Although both are quite minor, so maybe it doesn't matter that much.
+            // Sort by dApp ID, in ascending order (unstable sort should be faster, and stability is "guaranteed" due to lack of duplicated Ids).
             dapp_tiers.sort_unstable_by(|first, second| first.dapp_id.cmp(&second.dapp_id));
 
             // 5. Calculate rewards.
             let tier_rewards = tier_config
                 .reward_portion
                 .iter()
-                .map(|percent| *percent * dapp_reward_pool)
+                .zip(tier_config.slots_per_tier.iter())
+                .map(|(percent, slots)| {
+                    *percent * dapp_reward_pool / <u16 as Into<Balance>>::into(*slots)
```

Comment:
For any technically panicable math, let's have some comments to prove it's not possible to panick.

Reply:
This is a vulnerability, good catch!